### DISCUSSION TITLE
How to manage application and infrastructure deployment separately

### DISCUSSION THREAD

**Discussion by cyberjpb1** [fluxcd/flux2#3379] [Q&A] [answered]
How can I support the following case with Flux and a Helm Chart:

Suppose I have 2 environments: dev and prod

I create a version 1 of the chart which points to a Docker image A
I deploy chart version 1 in the dev environment
I deploy chart version 1 in the prod environment
I create a version 2 of the chart that points to a Docker image B
I deploy chart version 2 in the dev environment
I create a version 3 of the chart which changes the replicaCount
I deploy chart version 3 in dev (change=replicaCount + image B)
I deploy chart version 3 in prod (change=replicaCount + image B) but what I want is only the change of the replicaCount while keeping the current Docker image in prod, i.e. Docker image A

How can I have only replicaCount change in prod?

**Comment by cyberjpb1** [comment]
OK, I believe the solution is to create a new version of chart at deploy time and modify only the values.yaml file on the desired environment in the Azure DevOps deployment pipeline.

Am I in the right direction?

**Comment by ppawiggers** [comment, marked as answer]
I think the number of replicas should not be part of the chart. You should have a separate values file (or define it in the HelmRelease) that sets the number of replicas for the environment.

**Reply by kingdonb** [reply]
I think this is good advice. The chart (which is versioned and versions are immutable) includes the default values in `values.yaml`. If you change `values.yaml` then you must publish a new version of the chart.

If you "set" values through `spec.values` or through `valuesFrom` then you do not need to publish a new version of the chart, because the default hasn't changed – only been mutated by a consumer. If you set `valuesFiles` then things are a bit different, because now you've changed the defaults, but importantly *upstream* does not have to publish a new chart, or risk mutating other *downstream* environments by changing their defaults out from under them.
